libyul/Object.cpp

```diff
@@ -129,33 +141,60 @@ std::set<YulString> Object::qualifiedDataNames() const
 	return qualifiedNames;
 }
 
-std::vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
+void Object::visitPath(YulString _qualifiedName, std::function<bool(Object const*)> const& _visitor) const
 {
-	yulAssert(_qualifiedName != name, "");
-	yulAssert(subIndexByName.count(name) == 0, "");
-
 	if (boost::algorithm::starts_with(_qualifiedName.str(), name.str() + "."))
 		_qualifiedName = YulString{_qualifiedName.str().substr(name.str().length() + 1)};
 	yulAssert(!_qualifiedName.empty(), "");
 
 	std::vector<std::string> subObjectPathComponents;
 	boost::algorithm::split(subObjectPathComponents, _qualifiedName.str(), boost::is_any_of("."));
 
-	std::vector<size_t> path;
 	Object const* object = this;
 	for (std::string const& currentSubObjectName: subObjectPathComponents)
 	{
 		yulAssert(!currentSubObjectName.empty(), "");
-		auto subIndexIt = object->subIndexByName.find(YulString{currentSubObjectName});
-		yulAssert(
-			subIndexIt != object->subIndexByName.end(),
-			"Assembly object <" + _qualifiedName.str() + "> not found or does not contain code."
-		);
-		object = dynamic_cast<Object const*>(object->subObjects[subIndexIt->second].get());
-		yulAssert(object, "Assembly object <" + _qualifiedName.str() + "> not found or does not contain code.");
-		yulAssert(object->subId != std::numeric_limits<size_t>::max(), "");
-		path.push_back({object->subId});
+
+		YulString subObjectName = YulString{currentSubObjectName};
+		object = object->at(subObjectName);
+		if (object && _visitor(object))
+			break;
 	}
+}
 
+std::vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
+{
+	std::vector<size_t> path;
+	yulAssert(_qualifiedName != name, "");
+	yulAssert(subIndexByName.count(name) == 0, "");
+
+	this->visitPath(_qualifiedName, [&](Object const* _object) -> bool {
+		yulAssert(_object->subId != std::numeric_limits<size_t>::max(), "");
+		path.push_back(_object->subId);
+		return false;
+	});
```

Comment:
By the way `visitPath()` would be more idiomatic if it used iterators instead of callbacks. It would let us use ranges:

```
visitPath(_qualifiedName) | transform(objectToSubId) | to<vector>();
```

This would make most of these new helpers obsolete. I'd not do it in this PR thopugh. We want to be finally done with it :)